File: code/backend/src/main/java/com/taskflow/modules/whiteboard/service/WhiteboardService.java
package com.taskflow.modules.whiteboard.service;

import com.taskflow.modules.whiteboard.dto.CreateWhiteboardRequest;
import com.taskflow.modules.whiteboard.dto.SyncWhiteboardElementsRequest;
import com.taskflow.modules.whiteboard.dto.UpdateWhiteboardRequest;
import com.taskflow.modules.whiteboard.dto.WhiteboardDto;

import java.util.List;
import java.util.UUID;

public interface WhiteboardService {

    WhiteboardDto createWhiteboard(UUID userId, UUID workspaceId, CreateWhiteboardRequest request);

    List<WhiteboardDto> getWorkspaceWhiteboards(UUID userId, UUID workspaceId);

    WhiteboardDto getWhiteboardDetails(UUID userId, UUID whiteboardId);

    WhiteboardDto updateWhiteboard(UUID userId, UUID whiteboardId, UpdateWhiteboardRequest request);

    void deleteWhiteboard(UUID userId, UUID whiteboardId);

    WhiteboardDto syncElements(UUID userId, UUID whiteboardId, SyncWhiteboardElementsRequest request);
}

File: code/frontend/src/features/whiteboard/services/whiteboard-service.ts
import { apiClient } from '@/lib/api-client';
import type { ApiResponse } from '@/features/auth/types';
import type {
  WhiteboardDto,
  CreateWhiteboardPayload,
  UpdateWhiteboardPayload,
  SyncElementsPayload,
} from '../types';

export const whiteboardService = {
  getWorkspaceWhiteboards: async (workspaceId: string): Promise<WhiteboardDto[]> => {
    const res = await apiClient.get<ApiResponse<WhiteboardDto[]>>(
      `/workspaces/${workspaceId}/whiteboards`
    );
    return res.data.data;
  },

  createWhiteboard: async (workspaceId: string, data: CreateWhiteboardPayload): Promise<WhiteboardDto> => {
    const res = await apiClient.post<ApiResponse<WhiteboardDto>>(
      `/workspaces/${workspaceId}/whiteboards`,
      data
    );
    return res.data.data;
  },

  getWhiteboardDetails: async (whiteboardId: string): Promise<WhiteboardDto> => {
    const res = await apiClient.get<ApiResponse<WhiteboardDto>>(`/whiteboards/${whiteboardId}`);
    return res.data.data;
  },

  updateWhiteboard: async (whiteboardId: string, data: UpdateWhiteboardPayload): Promise<WhiteboardDto> => {
    const res = await apiClient.put<ApiResponse<WhiteboardDto>>(`/whiteboards/${whiteboardId}`, data);
    return res.data.data;
  },

  deleteWhiteboard: async (whiteboardId: string): Promise<void> => {
    await apiClient.delete<ApiResponse<void>>(`/whiteboards/${whiteboardId}`);
  },

  syncElements: async (whiteboardId: string, data: SyncElementsPayload): Promise<WhiteboardDto> => {
    const res = await apiClient.post<ApiResponse<WhiteboardDto>>(
      `/whiteboards/${whiteboardId}/elements/sync`,
      data
    );
    return res.data.data;
  },
};
